CLUSTERSCAN Command

nmvk opened the pull request:
Implemented CLUSTERSCAN command for topology-aware scanning

Unlike `SCAN` which is local to a single node, `CLUSTERSCAN` provides a
mechanism that helps clients iterate across slot boundaries and handles
`MOVED` redirections.

**Key details**

* Global cluster iteration via `fingerprint-{hashtag}-cursor`
* Scan one slot at a time
* Start the CLUSTERSCAN with 0
* SLOT argument for parallel scanning of multiple slots
* Re-use scanGenericCommand for the response

**Cursor format:** `fingerprint-{hashtag}-localcursor`
 - Fingerprint is a hash of the node's DB seed that identifies the
   current memory layout. On mismatch, scan restarts from cursor 0
   rather than returning an error.
 - Fingerprint 0 indicates a cross slot cursor (e.g., initial cursor
   or slot transition) where validation is skipped.
 - Hashtag encodes the target slot
 - Local cursor tracks position within the slot

**Usage:**

```
CLUSTERSCAN <cursor> [MATCH pattern] [COUNT count] [TYPE type] [SLOT number]
```

```
  CLUSTERSCAN 0                    # Start scanning from slot 0
  CLUSTERSCAN <cursor>             # Continue from cursor
  CLUSTERSCAN 0 SLOT 1000          # Start scanning specific slot
  CLUSTERSCAN <cursor> MATCH user:* COUNT 100
```

This closes https://github.com/valkey-io/valkey/issues/33

madolson commented on src/cluster.c:
```diff
@@ -1626,3 +1626,144 @@ void clusterCommandFlushslot(client *c) {
+/* -----------------------------------------------------------------------------
+ * CLUSTERSCAN Command
+ * -------------------------------------------------------------------------- */
+ static const char *getHashTagForSlot(int slot) {
```
```suggestion
 static const char *getHashTagForSlot(int slot) {
```

I think you can use `return (slot < CLUSTER_SLOTS) ? crc16_slot_table[slot] : NULL;` instead. You can see the implementation of VM_ClusterCanonicalKeyNameInSlot as well

madolson commented on src/cluster.c:
```diff
@@ -42,6 +42,9 @@
 #include "module.h"
 
+/* Include crc16_slot_table from crc16_slottable.h */
+extern const char crc16_slot_table[][4];
```
Let's put this in a header instead of just extern const'ing it. Either that or add a function.

madolson commented on src/cluster.c:
```diff
@@ -1628,3 +1631,116 @@ void clusterCommandFlushslot(client *c) {
+            if (input_slot < 0 || input_slot >= CLUSTER_SLOTS) {
+                addReplyError(c, "Invalid slot");
+                return;
+            }
```
getSlotOrReply covers this logic. We should also make sure slot is only provided once.

madolson commented on src/cluster.c:
```diff
@@ -1628,3 +1631,116 @@ void clusterCommandFlushslot(client *c) {
+            slot = 0;
+        }
+
+        sds new_cursor = sdscatprintf(sdsempty(), "0-{%s}-0", crc16_slot_table[slot]);
```
```suggestion
        sds new_cursor = sdscatfmt(sdsempty(), "0-{%S}-0", crc16_slot_table[slot]);
```
Should be faster.

madolson commented on src/commands/clusterscan.json:
```diff
@@ -0,0 +1,74 @@
+            "RESPONSE_POLICY:SPECIAL"
+        ],
+        "arguments": [
+            {
+                "name": "cursor",
+                "type": "string"
+            },
```
Use the sharded pubsub as an example. But if this is marked as not a key, then you don't need the custom getKey handler. Clients will also know to route this like a key.

nmvk replied:
Given the special case of `CLUSTERSCAN 0` marking it as `NOT_KEY` started treating 0 as a not special scenario and throwing MOVED redirects if node did not handle the slot that 0 gets hashed.

madolson replied:
I think that is OK? Is it that big of a deal if that case is routed to slot zero?

nmvk replied:
Thinking again, there is no meaningful advantage of current implementation over `NOT_KEY`.  The initial `CLUSTERSCAN 0` routes to arbitrary node, but the response is trivial. A client that wants to start scanning a slot directly can construct `0-{hashtag}-0` directly in first call and route to correct node and even get response. Updated to `NOT_KEY`

zuiderkwast replied:
We could mention one of the slot-0 words in the docs as an example to use as a start cursors, one being `0-{unwakened}-0`. There are many (actually not that many) words that hash to slot zero:

* unhidable
* unwakened
* lessors
* swathed
* urea
* ulcer
* interclass
* squatterarchy
* glopped
* Margret
* strong-mindedly
* Bondsville

nmvk replied:
`0-{squatterarchy}-0` is tempting. I can add `verbose-cursor-hashslot` config option in RC2 :D

zuiderkwast replied:
Yeah, use real words for each slot! 🤣 A great joke that users will not even see, since it's handled by a client lib under the hood...

I meant we can actually do this in the docs, to make the docs less boring than if we use `0-{06S}-0`.

nmvk replied:
Doc [PR](https://github.com/valkey-io/valkey-doc/pull/417) is now `swathed` with `flavor` instead of `06S` and `Qi`.

madolson commented on tests/unit/cluster/clusterscan.tcl:
```diff
@@ -0,0 +1,404 @@
+
+        # Version not 0
+        catch {R 0 clusterscan "1-{0}-0"} err
+        assert_match "*Invalid cursor*" $err
```
```suggestion
        assert_error "*Invalid cursor*" {R 0 clusterscan "1-{0}-0"}
```
This is a simpler one line of this.

madolson commented on src/cluster.c:
```diff
@@ -1628,3 +1631,116 @@ void clusterCommandFlushslot(client *c) {
+
+    /* Handle version */
+    token = strchr(p, '-');
+    if (!token) return C_ERR;
```
It's worth documenting why the version is needed somewhere, probably in the top level comment. The original point was so that during upgrades you could safely handle transparent re-directs.

madolson commented on tests/unit/cluster/clusterscan.tcl:
```diff
@@ -0,0 +1,404 @@
+                    break
+                } else {
+                    # Check if MOVED
+                    if {[string match "MOVED *" $res]} {
```
Ideally we should update the client so that it transparently handles everything. There is a `cluster.tcl` file that implements the cluster client.

madolson commented on src/crc16_slottable.h:
```diff
@@ -1,836 +1,11 @@
-"wu", "0TG", "4oi", "6ZJ" };
-
+typedef char crc16_alphastring[4];
+extern const crc16_alphastring crc16_slot_table[];
```
```suggestion
```

nmvk committed:
Implemented CLUSTERSCAN command for topology-aware scanning

Unlike `SCAN` which is local to a single node, `CLUSTERSCAN` provides a
mechanism that helps clients iterate across slot boundaries and handles
`MOVED` redirections.

**Key details**

* Global cluster iteration via `fingerprint-{hashtag}-cursor`
* Scan one slot at a time
* Start the CLUSTERSCAN with 0
* SLOT argument for parallel scanning of multiple slots
* Re-use scanGenericCommand for the response

**Cursor format:** `fingerprint-{hashtag}-localcursor`
 - Fingerprint is a hash of the node's DB seed that identifies the
   current memory layout. On mismatch, scan restarts from cursor 0
   rather than returning an error.
 - Fingerprint 0 indicates a cross slot cursor (e.g., initial cursor
   or slot transition) where validation is skipped.
 - Hashtag encodes the target slot
 - Local cursor tracks position within the slot

**Usage:**

CLUSTERSCAN <cursor> [MATCH pattern] [COUNT count] [TYPE type] [SLOT number]

  CLUSTERSCAN 0                    # Start scanning from slot 0
  CLUSTERSCAN <cursor>             # Continue from cursor
  CLUSTERSCAN 0 SLOT 1000          # Start scanning specific slot
  CLUSTERSCAN <cursor> MATCH user:* COUNT 100

This implements https://github.com/valkey-io/valkey/issues/33

Signed-off-by: nmvk <r@nmvk.com>

nmvk committed:
Testing the TLS Failure with bulk move

IO Error

Signed-off-by: nmvk <r@nmvk.com>

zuiderkwast commented on src/cluster.c:
```diff
@@ -1630,3 +1631,169 @@ void clusterCommandFlushslot(client *c) {
+        return;
+    }
+    /* Scan the slot using scanGenericCommand */
+    sds cursor_prefix = sdscatfmt(sdsempty(), "%u-{%s}-", fp, clusterGetSlotHashtag(slot));
```
The fingerprint can be up to 10 characters if we write it out in base 10. It's static, so it'd be nice if it'd be small. If we write it in hexadecimal, it fits in 8 chars (FFFFFFFF).

We could make it 6 chars if we do a base-64 variant, and if we reduce the fingerprint to 30 bytes, it fits in 5 chars (5 * 6 bits). It might be overkill though. 😄

<details>
<summary>Just for fun, here's a minimal base 64 encoding</summary>

It doesn't use `{ } -`. It's using a consecutive range of ascii characters starting from '0'.

```C
void encode(uint32_t num, char output[6]) {
    for (int i = 5; i >= 0; i--) {
        output[i] = '0' + (num & 0x3F);  // num % 64
        num >>= 6;  // num / 64
    }
}

uint32_t decode(const char input[6]) {
    uint32_t num = 0;
    for (int i = 0; i < 6; i++) {
        num = (num << 6) + (input[i] - '0');  // num * 64 + digit
    }
    return num;
}
```
</details>

madolson replied:
I agree, let's do base 64 encoding.

nmvk replied:
Thank you so much, I did not consider the encoding. I liked the base 64 style encoding.

zuiderkwast commented on src/db.c:
```diff
@@ -1219,6 +1223,13 @@ void scanGenericCommand(client *c, robj *o, unsigned long long cursor) {
+            if (slot == -1) {
+                addReplyError(c, "SLOT option is only valid for CLUSTERSCAN");
+                return;
+            }
```
For the other scan commands, SLOT is just an unknown argument. It doesn't need a dedicated error message IMO. We can fall through to the generic syntax error case below.

```suggestion
        } else if (slot != -1 && !strcasecmp(objectGetVal(c->argv[i]), "slot") && j >= 2) {
            /* SLOT is only valid for CLUSTERSCAN. */
```

zuiderkwast commented on src/cluster.c:
```diff
@@ -1630,3 +1631,169 @@ void clusterCommandFlushslot(client *c) {
+
+    /* Get the SLOT information if provided.*/
+    for (int i = 2; i < c->argc; i++) {
+        if (!strcasecmp(objectGetVal(c->argv[i]), "slot") && i + 1 < c->argc) {
```
Just searching for a "slot" argument can be wrong. If there'a a pattern that's exactly "slot", it would match in the wrong place.

    CLUSTERSCAN 0-{^_^}-0 MATCH slot SLOT 42

We should probably parse all the arguments together and only once. We could lift the parsing code out of scanGenericCommand to a separate function if you need to.

madolson commented on src/cluster.c:
```diff
@@ -1630,3 +1631,169 @@ void clusterCommandFlushslot(client *c) {
+    memcpy(&seed_high, seed + 8, 8);
+    uint64_t hash = seed_low ^ seed_high;
+
+    /* Tomas Wang's 64 bit integer hash. */
```
Minor, maybe put this in util.c, it's not implemented in 3 places.

madolson commented on src/cluster.c:
```diff
@@ -1630,3 +1631,169 @@ void clusterCommandFlushslot(client *c) {
+ *
+ * Returns a non zero 32-bit fingerprint. 0 is reserved for cross-node cursor */
+static uint32_t clusterScanFingerprint(void) {
+    uint8_t *seed = hashtableGetHashFunctionSeed();
```
Feels like this should be a static value, this won't change during the life time of the instance.

nmvk replied:
I updated it `static char cached_fp[7];` and caching

madolson commented on tests/unit/cluster/clusterscan.tcl:
```diff
@@ -0,0 +1,407 @@
+        
+        # Continue scanning until exhausted
+        set iterations 0
+        while {$cursor ne "0" && $iterations < 1000} {
```
Presumably also assert that iterations is < 1000.

madolson commented on tests/unit/cluster/clusterscan.tcl:
```diff
@@ -0,0 +1,407 @@
+        set max_loops 20000 ;# Safety break
+        
+        set iterations 0
+        while {$iterations < 20000} {
```
```suggestion
        while {$iterations < $max_loops} {
```

madolson commented on tests/unit/cluster/clusterscan.tcl:
```diff
@@ -0,0 +1,407 @@
+        foreach k $matched_keys {
+            assert_match "user:*" $k
+        }
+        assert {[llength $matched_keys] >= 2}
```
Since you did lsort unique, shouldn't this be exactly 2?

madolson commented on tests/unit/cluster/clusterscan.tcl:
```diff
@@ -0,0 +1,407 @@
+    }
+}
+
+# CLUSTERSCAN CLUSTERDOWN test - separate cluster to test unassigned slots
```
What exactly is this test verifying?

nmvk replied:
This test validates the `-CLUSTERDOWN` scenario when a slot is unassigned. We can't return MOVED since no node owns the slot. This check exists in the implementation to avoid crashes or silent failures. The test uses a separate cluster so `DELSLOTS` doesn't conflict with other tests.

madolson reviewed (commented):
Feels like it's getting closer.

nmvk committed:
Wang hash refactor, Base 64 style fp encoding, Static caching

Handle the PR comments related to Base64 style encoding, caching fp
fixing the argument parsing.

Signed-off-by: nmvk <r@nmvk.com>

nmvk committed:
Mark NOT_KEY for the ClusterScan

Mark as NOT_KEY this does make it easier to route,
though cluster 13907 would handle the clusterscan 0
it is possible to avoid this routing by using
hashtag this can be a client implementation detail.

Signed-off-by: nmvk <r@nmvk.com>

madolson commented on tests/unit/cluster/clusterscan.tcl:
```diff
@@ -0,0 +1,455 @@
+        set other_node [expr {1 - $slot0_owner}]
+        catch {R $other_node CLUSTER DELSLOTS 0}
+
+        after 200
```
```suggestion
        wait_for_cluster_state "fail"
```

madolson commented on tests/unit/cluster/clusterscan.tcl:
```diff
@@ -0,0 +1,455 @@
+        R $slot0_owner CLUSTER SETSLOT 0 NODE $target_id
+        R $target_node CLUSTER SETSLOT 0 NODE $target_id
+        
+        after 100
```
Waits are generally flaky with valgrind, so we try to not have fixed times. Some type of wait_for would be better.

madolson reviewed (approved):
Looks OK to me. Just some minor improvements to tests.

nmvk committed:
Fix the waits in the tests

Signed-off-by: nmvk <r@nmvk.com>

zuiderkwast commented on src/cluster.c:
```diff
@@ -1630,3 +1631,180 @@ void clusterCommandFlushslot(client *c) {
+    uint8_t *seed = hashtableGetHashFunctionSeed();
+    uint64_t seed_low, seed_high;
+    memcpy(&seed_low, seed, 8);
+    memcpy(&seed_high, seed + 8, 8);
```
Is it beneficial to avoid memcpy? I don't know if it matters, but I think this is equivalent:

```suggestion
    uint64_t *seed = (uint64_t *)hashtableGetHashFunctionSeed();
    uint64_t seed_low = seed[0], seed_high = seed[1];
```

nmvk replied:
Updated it since it is more readable.

zuiderkwast commented on src/module.c:
```diff
@@ -9578,7 +9578,7 @@ unsigned int VM_ClusterKeySlot(ValkeyModuleString *key) {
  * a valid slot. */
 const char *VM_ClusterCanonicalKeyNameInSlot(unsigned int slot) {
-    return (slot < CLUSTER_SLOTS) ? crc16_slot_table[slot] : NULL;
+    return (slot < CLUSTER_SLOTS) ? clusterGetSlotHashtag(slot) : NULL;
```
It doesn't have to be used as a hashtag. It can be the full name of a key too.

Not sure if it matters, but here's a suggestion:

```suggestion
    return (slot < CLUSTER_SLOTS) ? clusterKeyInSlot(slot) : NULL;
```

zuiderkwast commented on src/crc16_slottable.h:
```diff
@@ -1,836 +1,8 @@
- */
-typedef char crc16_alphastring[4];
-
-const crc16_alphastring crc16_slot_table[] = {
```
The point of this change is to avoid multiple copies of this array in the binary, or is there some other reason?

If that's the case, it's not strictly necessary to change it to a function. We could declare it as an `extern` global variable. (Defined in the .c file and declared extern in the .h file.) Then, we wouldn't need to change any of the code that accesses this.

nmvk replied:
Yes, it was to avoid the duplicate copies. Good point on extern being defined in header, no change in code that access it.

zuiderkwast reviewed (commented):
LGTM. I have a few minor ideas, but nothing that I feel strongly about.

nmvk committed:
Update the CRC function and avoid MEMCPY

Fixes to CRC changes

Signed-off-by: nmvk <r@nmvk.com>

zuiderkwast commented on src/cluster.c:
```diff
@@ -1630,3 +1631,177 @@ void clusterCommandFlushslot(client *c) {
+        }
+    }
+
+    /* Handle cursor "0" case. If slot information is provided we return
+     * the updated cursor to scan input slot, else scan from slot 0. */
+    if (strcmp(objectGetVal(c->argv[1]), "0") == 0) {
+        if (input_slot != -1) {
+            slot = input_slot;
+        } else {
+            slot = 0;
+        }
+
+        sds new_cursor = sdscatfmt(sdsempty(), "0-{%s}-0", crc16_slot_table[slot]);
+
+        addReplyArrayLen(c, 2);
+        addReplyBulkSds(c, new_cursor);
+        addReplyArrayLen(c, 0);
+        return;
```
If the current node owns slot 0 (or the `input_slot` if provided), can't we just start scanning directly instead of returning a new cursor first?

I'm thinking about clients who want to scan slots in parallel, one slot at a time. They'd be able to send the command to the correct node for the slot, but we can spare them the hassle of reverse-generating a word for each slot (like `0-{extrapolatory}-0` for slot 100) using commands like

    CLUSTERSCAN 0 SLOT 0
    CLUSTERSCAN 0 SLOT 1
    CLUSTERSCAN 0 SLOT 2

A client lib that provides a specialized parallel-cluster-scan feature wouldn't need to care about the keyspecs. It'd just generate the CLUSTERSCAN commands and send them to the correct nodes for each slot.

Seems like a simple quick path for this case:

```C
    if (getNodeBySlot(slot) == myself) {
        /* Handled by this node */
        cursor = 0;
    } else {
        sds new_cursor = sdscatfmt(sdsempty(), "0-{%s}-0", crc16_slot_table[slot]);

        addReplyArrayLen(c, 2);
        addReplyBulkSds(c, new_cursor);
        addReplyArrayLen(c, 0);
        return;
    }
```

nmvk replied:
The cursor has NOT_KEY so clients can't auto-route `CLUSTERSCAN 0 SLOT <n>` to the correct node, it would land on the node that `0` gets hashed. Hence we cannot start the scan directly. Only case would be if the node is same which is trivial.

zuiderkwast replied:
You mean slot("0") = 13907 and slot 0 are on the same shard? Yeah that's pretty unlikely. We don't need the special case for that.

So for cursor "0", all nodes redirect to the node that owns slot 13907 and this node will return a cursor like `0-{06S}-0` to start the scan then... Hm... Then I think we should document the possibility for clients to pick a smart string for the start cursor.

It's common that when clients receive a MOVED redirect, they think they'll need to update their slot mapping and issue CLUSTER SLOTS. ~~I suspect this NOT_KEY behavior and the reverse-hashtag-based cursors will just cause trouble for clients. It'd be easier and more direct for them if we'd just let the clients route this command by slot directly, I mean `CLUSTERSCAN whatever SLOT 42` should be sent to the node that owns slot 42. (I've already lost this fight, but I think we're making a mistake.)~~

zuiderkwast replied:
Oh, I guess I'm wrong again. A more complete scenario will can look like this:

1. Application code issues "CLUSTERSCAN 0 SLOT 42".
2. The client library isn't necessarily aware of CLUSTERSCAN but if it handles the keyspecs and the NOT_KEY flag, it will hash slot("0") = 13907 and send the command to the node that owns 13907.
3. Node 13907 replies with a new cursor like `0-{AUX}-0` (where slot("AUX") = 42).
4. Application sends command "CLUSTERSCAN 0-{AUX}-0 SLOT 42".
5. Client lib routes the command to the node that owns slot 42.
6. Node returns keys and new cursor still with the same hashtag inside.
7. Repeat 4-6 with new cursors until node returns cursor "0".

In this flow, there's no MOVED redirect involved. That's good. 👍

zuiderkwast commented on src/cluster.c:
```diff
@@ -1630,3 +1631,177 @@ void clusterCommandFlushslot(client *c) {
+    if (server.cluster->slots[slot] != getMyClusterNode()) {
+        clusterRedirectClient(c, server.cluster->slots[slot], slot, CLUSTER_REDIR_MOVED);
+        return;
+    }
```
Looks like this doesn't allow read from replicas. Do we want to allow it? If yes, it'd be something like this (untested):


```suggestion
    clusterNode *owner = server.cluster->slots[slot];
    clusterNode *myself = getMyClusterNode();
    if (owner != myself &&
        !(c->flag.readonly && clusterNodeIsReplica(myself) && owner == clusterNodeGetPrimary(myself)) {
        clusterRedirectClient(c, owner, slot, CLUSTER_REDIR_MOVED);
        return;
    }
```

nmvk replied:
Keeping this primary-only is my recommendation as it avoids the fingerprint mismatches. Given that a random hash seed is the default, primary and replica will have different hash seeds. Unlike conventional `GET` load balancing might not be a priority for `CLUSTERSCAN`  over better consistency by avoiding duplicates related to fingerprint mismatch. Final reason is we can change the behavior in future to support replicas but would not be able to restrict it.

madolson replied:
The fingerprints will generally match between primary and replica except for when there are upgrades. I also don't even think we need Viktor's recommendation to properly handle the routing, since it will be handled in the getKeysByCommand system. Or am I missing something?

nmvk replied:
Thanks @madolson I removed this check `NOT_KEY` handles this routing and this validation is not required here. Removed it

madolson replied:
I think this isn't really needed? Clients will always direct to primaries as needed and clients can opt-in to replica only behavior.

zuiderkwast replied:
> clients can opt-in to replica only behavior.

No, with this check, clients can't opt in to replica only behavior. CLUSTERSCAN will always redirect to the primary.

I'm fine with that though. The possibility that the replica seed is mismatching is a good callout.

zuiderkwast replied:
Ah, I see the last commit. Yeah, if we don't do an explicit redirect here, then I suppose it will work to read from replicas just like for other commands.

nmvk committed:
Remove the move direction

NOT_KEY handles this

Signed-off-by: nmvk <r@nmvk.com>

zuiderkwast commented on tests/unit/cluster/clusterscan.tcl:
```diff
@@ -0,0 +1,459 @@
+start_server {tags {external:skip} overrides {cluster-enabled no}} {
+    test "CLUSTERSCAN on CMD returns error" {
```
CMD is an AWS-internal acronym. It's not used outside of AWS. The rest of the world uses the term "standalone mode" or non-cluster mode or as in the error message "This instance has cluster support disabled".

```suggestion
    test "CLUSTERSCAN when not in cluster mode returns error" {
```

nmvk committed:
Update the test name for standalone mode

Update tests/unit/cluster/clusterscan.tcl

Co-authored-by: Viktor Söderqvist <viktor.soderqvist@est.tech>
Signed-off-by: Raghav <r@nmvk.com>

madolson commented on src/commands/clusterscan.json:
```diff
@@ -0,0 +1,92 @@
+            ]
+        }
+    }
+}
```
```suggestion
}

```
Minor comment to kickoff extra tests.

zuiderkwast replied:
@madolson This is not needed anymore after #3202 and the related changes. Just add the label and the tests will start.

Didn't it work?

madolson replied:
~They still didn't run actually, is it broken again?~

Nvm, I just got confused because we still skip some of the daily tests. I think it all ran as intended.

madolson committed:
Apply suggestion from @madolson

Signed-off-by: Madelyn Olson <madelyneolson@gmail.com>
